### PR TITLE
Upgrade bigquery common to 0.31.0, add impersonation support

### PR DESCRIPTION
**Pull request by yigress**

### CODE REVIEW

File: README-template.md [hunk 1]
```diff
@@ -412,6 +412,55 @@ Please note there are a few caveats:
...
+  implementation in the `bq.access.token.provider.fqcn` configuration property. The class must be
+  implemented in Java or other JVM languages such as Scala or Kotlin. It must either have a no-arg
+  constructor or a constructor accepting a single `java.util.String` parameter. This parameter can
+  be supplied using the `bq.access.token.provider.config` configuration property. If the property is
+  not set then the no-arg constructor will be called. The JAR containing the implementation class
+  should be on the cluster's classpath.
+* Define service account impersonation for specific users, specific groups, or for all
+  users by default using below properties:
+
+    - `bq.impersonation.service.account.for.user.<USER_NAME>` (not set by default)
```
Comment (functicons) [inline README-template.md:439]: Is this Linux user or Google account user?

File: README-template.md [hunk 2]
```diff
@@ -412,6 +412,55 @@ Please note there are a few caveats:
...
+
+      The service account to be impersonated for a specific user. You can specify multiple
+      properties using that pattern for multiple users.
+
+    - `bq.impersonation.service.account.for.group.<GROUP_NAME>` (not set by default)
```
Comment (functicons) [inline README-template.md:444]: ditto

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/integration/ImpersonationTests.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                      RuntimeException.class,
+                      () -> createManagedTable("abcd", HIVE_ALL_TYPES_TABLE_DDL));
+              JsonObject error =
+                  getErrorJsonFromHttpResponseException(
+                      exception,
+                      "com.google.auth.oauth2.ImpersonatedCredentials",
+                      "refreshAccessToken");
+              assertEquals(
+                  error.getAsJsonObject("error").get("message").getAsString(),
+                  "Not found; Gaia id not found for email " + sa);
```
Comment (functicons) [inline connector/src/test/java/com/google/cloud/hive/bigquery/connector/integration/ImpersonationTests.java:89]: Is there info about the user "bob" in the error?

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/integration/ImpersonationTests.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
```
Comment (functicons) [inline connector/src/test/java/com/google/cloud/hive/bigquery/connector/integration/ImpersonationTests.java:116]: Can we show group info here?

File: README-template.md (hunk 1 above)
Comment (yigress) [reply README-template.md:439]: added comment to clarify that it is for the linux user that runs the hive query

File: README-template.md (hunk 2 above)
Comment (yigress) [reply README-template.md:444]: ditto

File: connector/src/test/java/com/google/cloud/hive/bigquery/connector/integration/ImpersonationTests.java (hunk 4 above)
Comment (yigress) [reply connector/src/test/java/com/google/cloud/hive/bigquery/connector/integration/ImpersonationTests.java:116]: this error is at config initiation time to check the service account, the error message contains the sa, the linux user info that actually runs the job is not available here.

**Pull request merged**